MPP-3390: CircleCI config refactor

jwhitlock opened the pull request:
While working on MPP-3390, I found that the current `.circleci/config.yaml` was fighting my changes. This makes a number of changes to the configuration.

*Update 1*: I added `.circleci/python_job.bash` to house the bash functions. I also updated the "django 4.2" job to "django 5.0", and added "python with postgres [15.5 / 16.1]" tests

* Add comments linking to the relevant CircleCI documentation for orbs, docker images, and configuration references.
* Use YAML anchors and aliases to repeat the docker authentication and tag filters
* In `python_job`, call a bash function rather than pass the command as a job parameter
* Expand `python_job` to replace `build_email_tracker_list_job`
* Expand `python_job` to replace `python_test_postgres`

More changes are coming once PR #4174 and PR #4175 are deployed to production, and I can update how the migrations tests run.

## How to test:

- [ ] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [ ] The "build_frontend" job runs in CircleCI. This is similar to the job with the same name in main. The "Build Code" step builds the Next.js code without errors. Warnings are similar to the build step on main. The job on main has an extra "Install Node.js 18" step, which updates to Node v18.19.0 or later.
- [ ] The "build email tracker lists" job runs in CircleCI. This is similar to the job with the same name in main. The "Run build email tracker lists" builds the two lists an copies them to `/tmp/workspace/email-trackers`. This is similar to the "build_email_tracker_list_job" on main, combining steps "Build level one email tracker list" (with less warnings), "Build level two email tracker list" (with less warnings), and "Copy email tracker lists to persisted workspace".
- [ ] The "build_test_backend" job runs in CircleCI. This is nearly identical "build_test_backend" job in main. The main job includes saving the docker image, the PR job does not. The job runs 910 tests in both. The "Test Code" step shows "895 passed, 15 skipped, 2 warnings" in both. 
- [ ] The "black style check" job runs in CircleCI. This is similar to the job with the same name in main. The "Run black" step is similar to the "black --check ." step in the main branch, showing "203 files would be left unchanged."
- [ ] The "mypy type check" job runs in CircleCI. This is similar to the job with the same name in main. Both show 1 Tests ran. The "Run mypy" step is similar to the "mypy --no-incremental --junit-xml job-results/mypy.xml ." step in main, showing "Success: no issues found in 203 source files". 
- [ ] The "mypy strict type check" job runs in CircleCI. This is similar to the job with the same name in main. Both show 1 Tests ran. The "Run mypy (failure allowed)" step is similar to the "mypy --strict --no-incremental --junit-xml job-results/mypy.xml . (failure allowed)" step in main, showing "Found 1726 errors in 88 files (checked 203 source files)".
- [ ] The "pytest with current postgres" job runs in CircleCI. This is similar to the "python_test_postgres" job in main. Both show 910 tests ran. The "Run pytest" step is similar to the "Run tests" step in main. This step shows "895 passed, 15 skipped, 5 warnings" in this branch, and "172 warnings" in main. The warning reduction is due to creating the `staticfiles` directory in the new code.
- [ ] The "sqlite migrations test" job runs in CircleCI. This is similar to the job with the same name in main. Both show 910 tests ran. The "Run pytest" step is similar to the "pytest --maxfail=3 --junit-xml=job-results/pytest-migrations.xml ." step in main. This step shows "895 passed, 15 skipped, 5 warnings" in this branch, and "2 warnings" in main. The extra warnings are for the "vobject" module. *I'm not sure why these are appearing more consistently with the new configuration.*
- [ ] The "postgres migrations test" job runs in CircleCI. This is similar to the job with the same name in main. Both show 910 tests ran. The "Run pytest" step is similar to the "Run tests" step in main. This step' shows "895 passed, 15 skipped, 5 warnings" in this branch, and "172 warnings" in main. The warning reduction is due to creating the `staticfiles` directory in the new code.
- [ ] The "python test phones disabled" job runs in CircleCI. This is similar to the job with the same name in main. Both show 910 tests ran. The "Run pytest" step is similar to the "PHONES_ENABLED=0 IQ_ENABLED=0 pytest --junit-xml=job-results/pytest-phones-disabled.xml ." step in main. Both show "671 passed, 239 skipped, 2 warnings".
- [ ] The ~"django 4.2 test"~ "django 5.0 test" job runs in CircleCI. This is similar to the "django 4.2 test" job in main. Both show 910 tests ran. The "Run pytest (failure allowed)" step is similar to the "pip install --pre --upgrade "Django<5"; pytest --maxfail=3 --junit-xml=job-results/pytest-django-4-2.xml . (failure allowed)" step in main. This branch shows "895 passed, 15 skipped, 4 warnings", and main shows "2 warnings". The extra warnings are for vobject.
- [ ] The "python test iq enabled" job runs in CircleCI. This is similar to the job with the same name in main. Both show 910 tests ran. The "Run pytest" step is similar to the "PHONES_ENABLED=1 IQ_ENABLED=1 pytest --junit-xml=job-results/pytest-iq-enabled.xml ." in main. This branch's step shows "908 passed, 2 skipped, 5 warnings", and main shows "2 warnings". The extra warnings are for the vobject module.
- [ ] The "test frontend" job runs in CircleCI. This is similar to the job with the same name in main. Both show 205 tests ran. The "Test Code" step both shows "Tests:   3 skipped, 202 passed, 205 total".
- [ ] The "convert_frontend_coverage" job runs in CircleCI. This is identical to the job with the same name in main, running the same steps.
- [ ] The ""upload_coverage" job runs in CircleCI. This is identical to the job with the same name in main, running the same steps.
- [ ] The "deploy" job does not run in CircleCI. It will run when this PR is merged to main. There are no significant changes to this job.
- [ ] The "heroku_deploy" job does not run in CircleCI. It will run when this PR is merged to main. There are no significant changes to this job.

## More Details

### Use YAML anchors and aliases

One of the features of YAML is [anchors and aliases](https://batalin.dev/posts/yaml-anchors-and-aliases/), which allow removing some duplication. For example, `&default_auth` declares an anchor for the map of username and password:

```yaml
executors:
    docker:
      - image: docker:<<parameters.docker-version>><<parameters.variant>>
        auth: &default_auth
          username: $DOCKER_USER
          password: $DOCKER_PASS
```

Later, this can be used as an aliases `*default_auth` to repeat the same `auth:` map:

```yaml
executors:
  ...
  python:
    # https://circleci.com/developer/images/image/cimg/python
    docker:
      - image: cimg/python:<<parameters.python-version>><<parameters.variant>>
        auth: *default_auth
```

A similar technique is used in the `workflows` section at the end of the file, to define the anchor `&default_filters` for our filter that runs tests on tags, and to use the alias `*default_filters` on the many tests that have the same filter. This helps the "deploy" and "heroku_deploy" jobs stand out as having different filters.

It is useful to see these expanded. I use `yq` for this:

```sh
brew install yq
yq -p yaml -o json .circleci/config.yml | yq -p json -o yaml | less 
```
This:

* Installs `yq` with Homebrew. [yq](https://mikefarah.gitbook.io/yq/) is like [jq](https://jqlang.github.io/jq/) for YAML.
* Use `yq` to convert `config.yml` to JSON, expanding the aliases
* Use `yq` to convert back to YAML
* Pipe to `less` so you can browse the YAML.

### Convert to executors
I converted jobs to use an `executor`. For example, the node jobs were:

```yaml
jobs:
  build_frontend:
    docker:
      - image: cimg/node:18.18
        auth:
          username: $DOCKER_USER
          password: $DOCKER_PASS
    steps:
       ...
  test_frontend:
    docker:
      - image: cimg/node:18.18
        auth:
          username: $DOCKER_USER
          password: $DOCKER_PASS
    steps:
        ...
```
and are now:

```yaml
executors:
  node:
    # https://circleci.com/developer/images/image/cimg/node
    docker:
      - image: cimg/node:<<parameters.node_version>>
        auth: *default_auth
    parameters:
      node_version:
        description: "version tag"
        default: "18.18"
        type: string
    ...

jobs:
  build_frontend:
    executor: node
    steps:
       ...

  test_frontend:
    executor: node
    steps:
       ...
```

This removes some duplication, and makes a single place to update the node version. It also enables using a [matrix](https://circleci.com/docs/reusing-config/#example-of-using-an-executor-declared-in-configyml-with-matrix-jobs) to run jobs like `build_frontend` on the current and next node versions, via the `node_version` parameter. The new jobs "python with postgres 15.5" and "python with postgres 16.1" are matrix jobs.

### Commands as bash functions

The `python_job` handles all jobs that run under the Python image. It handles this flexibility with a bunch of parameters with defaults. For example, the default executor is 'python', but 'executor: python_with_postgres' can be used to also spin up a PostgreSQL server.

It uses bash functions declared in the new file `.circleci/python_job.bash`. To configure these functions, CircleCI parameters are turned into Bash variables. These variables and `.circleci/python_job.bash` are appended to the environment so they are available for other steps.

jwhitlock commented:
I was pretty tired of CircleCI config after a couple days of incremental testing. I'm refreshed after a weekend.

I am curious if it is more helpful to have the bash functions in the config file, or if it would be better to move that to its own file. For example. there could be a `.circleci/relay_scripts.bash` file, and a line `echo 'source .cirlceci/relay_scripts.bash' >> $BASH_RC`. Then the CircleCI steps could just call the functions it sets up. This would replace a lot of "Section 2" of the `python_job`.

groovecoder commented:
> I was pretty tired of CircleCI config after a couple days of incremental testing. I'm refreshed after a weekend.
> 
> I am curious if it is more helpful to have the bash functions in the config file, or if it would be better to move that to its own file. For example. there could be a `.circleci/relay_scripts.bash` file, and a line `echo 'source .cirlceci/relay_scripts.bash' >> $BASH_RC`. Then the CircleCI steps could just call the functions it sets up. This would replace a lot of "Section 2" of the `python_job`.

Separate scripts seems helpful. [FXA has a `_scripts/` dir](https://github.com/mozilla/fxa/tree/main/_scripts) full of scripts that can be called from a number of places.

jwhitlock commented:
After several force-pushes, I have a working CircleCI configuration again. The changes:

* Most Bash functions are in the new file `.circleci/python_job.bash`. I don't think it is general enough for a script directory. It greatly simplifies the `python_job` definition.
* I changed the Django 4.2 job to a Django 5.0 job, since that was just released
* I added tests for PostgreSQL 15.5 and 16.1, because we've been thinking about upgrading, and to get matrix jobs working
* I changed some parameter names to use underscores, because my tab-completion doesn't work with dashes.

jwhitlock commented:
I've rebased, and changed the migrations test to test the production code and requirements against the branch migrations. Previously, the branch requirements were used, not the production requirements, making the migrations tests different from the mixed production push environment. Specifically, the branch may include updates to requirements that have their own migrations (django-allauth 😠)

jwhitlock commented:
Rebased on main

say-yawn commented on .circleci/config.yml:
```diff
@@ -340,11 +417,89 @@ jobs:
+            # set -x  # Uncomment to debug
+            TMP_ENV=$(mktemp)
+
+            if [ "<<# parameters.allow_fail >>allow_fail<</ parameters.allow_fail >>" == "allow_fail" ]
```
(suggestion, non-blocking) Add comment on the [mustache conditionals](https://support.circleci.com/hc/en-us/articles/4417604103835-Using-Mustache-Conditionals-in-Config-File) and/or consider using `steps` and `condition`.

jwhitlock replied:
The main documentation mentions conditions, and the "mustache" syntax is barely mentioned. However, it is a big change to switch to conditions, and it will take some experimentation to make it look decent. I think I'll do the small thing, adding some comments, and tackle the re-write in a new PR, so we can decide if it is an improvement.

say-yawn reviewed (approved):
Changes LGTM. Thanks @jwhitlock for cleaning the CircleCI config and adding more tests to ensure the code changes are being tested holistically. I made a non-blocking suggestion about using the mustache conditionals.
